### PR TITLE
Password complexity rules

### PR DESCRIPTION
**Pull request by evillique**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add password complexity rules and checks for creating a new user.

### CODE REVIEW

File: programs/client/Client.cpp [hunk 1]
```diff
@@ -180,6 +182,64 @@ std::vector<String> Client::loadWarningMessages()
 }
 
 
+/// Make query to get all server password complexity rules
+void Client::loadPasswordComplexityRules()
+{
+    /// Older server versions cannot execute the query loading password complexity rules.
+    constexpr UInt64 min_server_revision_to_load_warnings = DBMS_MIN_PROTOCOL_VERSION_WITH_VIEW_IF_PERMITTED;
```
Comment (nikitamikhaylov) [inline programs/client/Client.cpp:189]: `min_server_revision_to_load_complexity_rules`

File: src/Interpreters/Access/InterpreterCreateUserQuery.cpp [hunk 2]
```diff
@@ -108,6 +108,12 @@ BlockIO InterpreterCreateUserQuery::execute()
         throw Exception(ErrorCodes::BAD_ARGUMENTS,
             "Authentication type NO_PASSWORD must be explicitly specified, check the setting allow_implicit_no_password in the server configuration");
 
+    if (!query.attach && query.temporary_password_for_checks)
```
Comment (nikitamikhaylov) [inline src/Interpreters/Access/InterpreterCreateUserQuery.cpp:111]: So we won't check password complexity for existing users? This should be backward compatible then

File: programs/client/Client.cpp [hunk 3]
```diff
@@ -258,6 +318,16 @@ try
...
         showWarnings();
 
+    /// Load password complexity rules
+    try
+    {
+        loadPasswordComplexityRules();
+    }
+    catch (...)
+    {
+        /// Ignore exception
```
Comment (nikitamikhaylov) [inline programs/client/Client.cpp:328]: We need at least write this exception to logs

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Interpreters/Access/InterpreterCreateUserQuery.cpp (hunk 2 above)
Comment (evillique) [reply src/Interpreters/Access/InterpreterCreateUserQuery.cpp:111]: Yes, these will only be applied to the new passwords

File: programs/client/Client.cpp (hunk 3 above)
Comment (nikitamikhaylov) [reply programs/client/Client.cpp:328]: Oh, forgot that we are on the client side... Can we simply rethrow an exception?

**Comment by evillique** [comment]
I slightly remade the implementation, so instead of getting the password rules with `SELECT` query from `system.passwords_rules`, we now get them in the initial HELLO message.

It now fetches the rules faster and saves us 1 RTT which affects workloads where clickhouse-client is consistently run in non-interactive mode and prevents clickhouse-client startup failures when the `max_concurrent_queries` is met.

**Review by nikitamikhaylov — APPROVED** [review]

File: src/Server/TCPHandler.cpp [hunk 4]
```diff
@@ -1193,6 +1194,17 @@ void TCPHandler::sendHello()
...
+    {
+        auto rules = server.context()->getAccessControl().getPasswordComplexityRules();
+
+        writeVarUInt(rules.size(), *out);
+        for (const auto & [original_pattern, exception_message] : rules)
+        {
+            writeStringBinary(original_pattern, *out);
+            writeStringBinary(exception_message, *out);
+        }
+    }
```
Comment (vitlibar) [inline src/Server/TCPHandler.cpp:1207]: So the password complexity rules are going to work only for TCP protocol?

Comment (evillique) [reply src/Server/TCPHandler.cpp:1207]: It seems like we only need to add complexity rules handling to clickhouse-client and therefore to our native TCP protocol. In all other cases (at least as I'm aware) we send the original password to the server, and therefore the server will check passwords without any problems.

Comment (vitlibar) [reply src/Server/TCPHandler.cpp:1207]: I got it. A test for http protocol would be useful

**Pull request merged**